File: backend/app/services/helpers/mailer_history.py
"""Normalize ``leads.mailer_history`` JSONB into a readable row list.

Handles legacy free-text strings (HubSpot/import era), OLC dict entries, and
mixed arrays. Canonical for API serializers (command-center payload); the FE
prefers ``mailer_history_summary`` from the API when present.
"""
from __future__ import annotations

import re
from datetime import datetime
from typing import Any

# Trailing date in legacy strings like "Boyfriend, OLM, Blue,  6/21/2024"
_LEGACY_DATE_RE = re.compile(
    r'(?P<label>.*?),\s*(?P<date>\d{1,2}/\d{1,2}/\d{2,4})\s*$',
)


def _as_entries(raw: Any) -> list[Any]:
    if raw is None or raw == '' or raw == []:
        return []
    if isinstance(raw, list):
        return list(raw)
    return [raw]


def parse_mailer_sent_at(value: Any) -> datetime | None:
    """Parse ISO or US slash dates for ordering; None when unparseable."""
    if value is None:
        return None
    text = str(value).strip()
    if not text:
        return None
    if text.endswith('Z'):
        text = text[:-1] + '+00:00'
    try:
        return datetime.fromisoformat(text)
    except ValueError:
        pass
    for fmt in ('%m/%d/%Y', '%m/%d/%y', '%Y-%m-%d', '%Y/%m/%d'):
        try:
            return datetime.strptime(text, fmt)
        except ValueError:
            continue
    return None


def normalize_mailer_history(raw: Any) -> list[dict[str, Any]]:
    """Return stable mail-history rows for UI.

    Each row:
      id, sent_at, label, creative, template_name, campaign_id,
      olc_order_id, address_feedback, cancelled, source
    """
    rows: list[dict[str, Any]] = []
    for idx, entry in enumerate(_as_entries(raw)):
        row = _normalize_one(entry, idx)
        if row is not None:
            rows.append(row)
    return rows


def _normalize_one(entry: Any, idx: int) -> dict[str, Any] | None:
    if entry is None or entry == '':
        return None

    if isinstance(entry, dict):
        sent_at = entry.get('sent_at')
        template_name = entry.get('template_name')
        creative = entry.get('creative')
        label_parts = [p for p in (template_name, creative) if p]
        label = ', '.join(str(p) for p in label_parts) if label_parts else None
        if not label and entry.get('olc_order_id'):
            label = f"OLC order {entry.get('olc_order_id')}"
        if not label and entry.get('campaign_id') is not None:
            label = f"Campaign {entry.get('campaign_id')}"
        if not label and entry.get('address_feedback'):
            label = f"Address feedback: {entry.get('address_feedback')}"
        if not label:
            label = 'Mailer'
        source = 'olc' if (
            entry.get('campaign_id') is not None or entry.get('olc_order_id')
        ) else 'imported'
        return {
            'id': f'mail-{idx}',
            'sent_at': sent_at if sent_at is None else str(sent_at),
            'label': label,
            'creative': creative,
            'template_name': template_name,
            'campaign_id': entry.get('campaign_id'),
            'olc_order_id': entry.get('olc_order_id'),
            'address_feedback': entry.get('address_feedback'),
            'cancelled': bool(entry.get('cancelled')),
            'source': source,
        }

    text = str(entry).strip()
    if not text:
        return None
    match = _LEGACY_DATE_RE.match(text)
    if match:
        return {
            'id': f'mail-{idx}',
            'sent_at': match.group('date'),
            'label': match.group('label').strip().rstrip(','),
            'creative': None,
            'template_name': None,
            'campaign_id': None,
            'olc_order_id': None,
            'address_feedback': None,
            'cancelled': False,
            'source': 'imported',
        }
    return {
        'id': f'mail-{idx}',
        'sent_at': None,
        'label': text,
        'creative': None,
        'template_name': None,
        'campaign_id': None,
        'olc_order_id': None,
        'address_feedback': None,
        'cancelled': False,
        'source': 'imported',
    }


def _last_sent_from_rows(rows: list[dict[str, Any]]) -> Any:
    """Date-aware last-sent lookup shared by summary + consolidate."""
    last_sent = None
    last_dt: datetime | None = None
    for row in rows:
        sent = row.get('sent_at')
        if not sent:
            continue
        parsed = parse_mailer_sent_at(sent)
        if parsed is not None:
            if last_dt is None or parsed > last_dt:
                last_dt = parsed
                last_sent = sent
        elif last_sent is None:
            last_sent = str(sent)
    return last_sent


def mailer_history_summary(raw: Any) -> dict[str, Any]:
    """Count + last sent_at for summary chips (date-aware ordering)."""
    rows = normalize_mailer_history(raw)
    return {
        'count': len(rows),
        'last_sent_at': _last_sent_from_rows(rows),
        'rows': rows,
    }


def _present_identifier(value: Any) -> str | None:
    if value is None:
        return None
    text = str(value).strip()
    return text or None


def _dedupe_keys(row: dict[str, Any]) -> set[tuple[str, str | None, str | None]]:
    """Identities for a mail-history row.

    OLC-backed rows can arrive with only one of campaign_id / olc_order_id, or
    with numeric IDs represented as strings. Treat any shared normalized OLC
    identifier as the same send; fall back to text identity for legacy rows.
    """
    campaign_id = row.get('campaign_id')
    olc_order_id = row.get('olc_order_id')
    keys: set[tuple[str, str | None, str | None]] = set()
    normalized_campaign_id = _present_identifier(campaign_id)
    normalized_olc_order_id = _present_identifier(olc_order_id)
    if normalized_campaign_id is not None:
        keys.add(('olc-campaign', normalized_campaign_id, None))
    if normalized_olc_order_id is not None:
        keys.add(('olc-order', normalized_olc_order_id, None))
    if keys:
        return keys
    return {('text', _present_identifier(row.get('sent_at')), _present_identifier(row.get('label')))}


def _timeline_mail_sent_rows(lead: Any) -> list[dict[str, Any]]:
    """Normalize ``mail_sent`` LeadTimelineEntry rows into the mail-history shape."""
    lead_id = getattr(lead, 'id', None)
    if not lead_id:
        return []
    from app.models.lead_timeline_entry import LeadTimelineEntry

    entries = (
        LeadTimelineEntry.query.filter(
            LeadTimelineEntry.lead_id == lead_id,
            LeadTimelineEntry.event_type == 'mail_sent',
            LeadTimelineEntry.is_deleted.is_(False),
        )
        .order_by(LeadTimelineEntry.occurred_at.asc())
        .all()
    )
    rows: list[dict[str, Any]] = []
    for entry in entries:
        metadata = entry.event_metadata if isinstance(entry.event_metadata, dict) else {}
        campaign_id = metadata.get('campaign_id')
        olc_order_id = metadata.get('olc_order_id')
        template_name = metadata.get('template_name')
        creative = metadata.get('creative')
        label_parts = [p for p in (template_name, creative) if p]
        label = ', '.join(str(p) for p in label_parts) if label_parts else None
        if not label and olc_order_id:
            label = f'OLC order {olc_order_id}'
        if not label and campaign_id is not None:
            label = f'Campaign {campaign_id}'
        if not label:
            label = entry.summary or 'Mailer sent'
        rows.append({
            'id': f'timeline-{entry.id}',
            'sent_at': entry.occurred_at.isoformat() if entry.occurred_at else None,
            'label': label,
            'creative': creative,
            'template_name': template_name,
            'campaign_id': campaign_id,
            'olc_order_id': olc_order_id,
            'address_feedback': metadata.get('address_feedback'),
            'cancelled': bool(metadata.get('cancelled')),
            'source': 'timeline',
        })
    return rows


def _heal_mailer_history_gaps(lead: Any, healed_entries: list[dict[str, Any]]) -> None:
    """Append OLC dicts missing from ``lead.mailer_history`` and flag_modified.

    Mirrors the canonical JSONB-append pattern used by mail_campaign_service
    (``_stamp_address_feedback`` / ``_stamp_silent_omit``): read the current
    list, append plain dicts, reassign, then ``flag_modified``.
    """
    from sqlalchemy.orm.attributes import flag_modified

    raw = getattr(lead, 'mailer_history', None)
    if isinstance(raw, list):
        history = list(raw)
    elif raw:
        history = [raw]
    else:
        history = []
    for entry in healed_entries:
        history.append({
            k: v for k, v in entry.items()
            if v is not None
        })
    lead.mailer_history = history
    flag_modified(lead, 'mailer_history')


def consolidate_mailer_history(lead: Any, *, heal: bool = True) -> dict[str, Any]:
    """Union ``lead.mailer_history`` JSONB with ``mail_sent`` timeline rows.

    The OLC submit pipeline occasionally drops a campaign send from the
    ``mailer_history`` JSONB column (silent omit) while the ``mail_sent``
    LeadTimelineEntry still recorded it — see the 10305-class bug where a
    lead's import-string history undercounted mailers actually sent. This
    unions both sources, dedupes by any shared normalized OLC identifier
    (falling back to ``(sent_at, label)`` for legacy free-text rows),
    and — when ``heal`` is True — appends any timeline-only mailer as a plain
    OLC dict onto ``lead.mailer_history`` so future reads see it directly
    from the JSONB column without needing this union again.

    There is no separate mail-campaign/lead join table today (a
    ``MailCampaign`` only stores lead ids in JSON tracking columns), so the
    two sources unioned here — JSONB history and timeline ``mail_sent``
    entries — are the complete set of "easily available" sources.

    Returns the same shape as :func:`mailer_history_summary` plus
    ``healed_count`` (rows newly appended to the JSONB column this call).
    """
    jsonb_rows = normalize_mailer_history(getattr(lead, 'mailer_history', None))
    seen_keys: set[tuple[str, str | None, str | None]] = set()
    for row in jsonb_rows:
        seen_keys.update(_dedupe_keys(row))

    merged_rows = list(jsonb_rows)
    healed_entries: list[dict[str, Any]] = []
    for row in _timeline_mail_sent_rows(lead):
        keys = _dedupe_keys(row)
        if keys & seen_keys:
            continue
        seen_keys.update(keys)
        merged_rows.append(row)
        healed_entries.append({
            'sent_at': row.get('sent_at'),
            'template_name': row.get('template_name'),
            'creative': row.get('creative'),
            'campaign_id': row.get('campaign_id'),
            'olc_order_id': row.get('olc_order_id'),
            'address_feedback': row.get('address_feedback'),
            'cancelled': row.get('cancelled') or False,
        })

    if heal and healed_entries:
        _heal_mailer_history_gaps(lead, healed_entries)

    return {
        'count': len(merged_rows),
        'last_sent_at': _last_sent_from_rows(merged_rows),
        'rows': merged_rows,
        'healed_count': len(healed_entries),
    }

File: frontend/src/utils/mailerHistory.ts
/**
 * Normalize leads.mailer_history (legacy string | JSON array | mixed) for UI.
 *
 * Prefer `mailer_history_summary` from the command-center API when present
 * (backend is the canonical normalizer). This module is the FE fallback for
 * raw `mailer_history` only.
 */

export type MailerHistorySource = 'olc' | 'imported' | 'timeline'

export interface MailerHistoryRow {
  id: string
  sent_at: string | null
  label: string
  creative: string | null
  template_name: string | null
  campaign_id: number | null
  olc_order_id: string | null
  address_feedback: string | null
  address_failure_reason?: string | null
  olc_silent_omit?: boolean
  cancelled: boolean
  source: MailerHistorySource
}

export interface MailerHistorySummary {
  count: number
  last_sent_at: string | null
  healed_count?: number
  rows: MailerHistoryRow[]
}

const LEGACY_DATE_RE = /^(?<label>.*?),\s*(?<date>\d{1,2}\/\d{1,2}\/\d{2,4})\s*$/
const DATE_ONLY_SLASH_RE = /^(\d{1,2})\/(\d{1,2})\/(\d{2,4})$/
const DATE_ONLY_ISO_RE = /^(\d{4})-(\d{2})-(\d{2})$/
const NAIVE_ISO_DATETIME_RE = /^\d{4}-\d{2}-\d{2}T/
const HAS_TZ_RE = /(Z|[+-]\d{2}:?\d{2})$/i

/** Product timestamps display in US Central (America/Chicago). */
export const MAILER_DISPLAY_TIME_ZONE = 'America/Chicago'

const CENTRAL_DATE_TIME_OPTS: Intl.DateTimeFormatOptions = {
  timeZone: MAILER_DISPLAY_TIME_ZONE,
  month: 'short',
  day: 'numeric',
  year: 'numeric',
  hour: 'numeric',
  minute: '2-digit',
  timeZoneName: 'short',
}

function asEntries(raw: unknown): unknown[] {
  if (raw == null || raw === '' || (Array.isArray(raw) && raw.length === 0)) {
    return []
  }
  if (Array.isArray(raw)) return [...raw]
  return [raw]
}

/** Parse ISO or US slash dates for last-sent ordering. Naive ISO datetimes are UTC. */
export function parseMailerSentAt(value: unknown): Date | null {
  if (value == null) return null
  const text = String(value).trim()
  if (!text) return null
  const slash = DATE_ONLY_SLASH_RE.exec(text)
  if (slash) {
    const month = Number(slash[1])
    const day = Number(slash[2])
    let year = Number(slash[3])
    if (year < 100) year += 2000
    const d = new Date(year, month - 1, day)
    if (
      Number.isNaN(d.getTime())
      || d.getFullYear() !== year
      || d.getMonth() !== month - 1
      || d.getDate() !== day
    ) {
      return null
    }
    return d
  }
  const isoText =
    NAIVE_ISO_DATETIME_RE.test(text) && !HAS_TZ_RE.test(text) ? `${text}Z` : text
  const ms = Date.parse(isoText)
  if (!Number.isNaN(ms)) return new Date(ms)
  return null
}

function formatCalendarDay(year: number, month: number, day: number): string {
  const utc = new Date(Date.UTC(year, month - 1, day))
  if (
    Number.isNaN(utc.getTime())
    || utc.getUTCFullYear() !== year
    || utc.getUTCMonth() !== month - 1
    || utc.getUTCDate() !== day
  ) {
    return '—'
  }
  return utc.toLocaleDateString('en-US', {
    timeZone: 'UTC',
    month: 'short',
    day: 'numeric',
    year: 'numeric',
  })
}

/**
 * Readable mail-history timestamp in US Central Time.
 * Date-only values stay calendar dates (no invented clock time / TZ shift).
 */
export function formatMailerSentAtDisplay(value: string | null | undefined): string {
  if (value == null) return '—'
  const text = String(value).trim()
  if (!text) return '—'

  const slash = DATE_ONLY_SLASH_RE.exec(text)
  if (slash) {
    let year = Number(slash[3])
    if (year < 100) year += 2000
    return formatCalendarDay(year, Number(slash[1]), Number(slash[2]))
  }

  const isoDate = DATE_ONLY_ISO_RE.exec(text)
  if (isoDate) {
    return formatCalendarDay(
      Number(isoDate[1]),
      Number(isoDate[2]),
      Number(isoDate[3]),
    )
  }

  const parsed = parseMailerSentAt(text)
  if (!parsed) return '—'
  return parsed.toLocaleString('en-US', CENTRAL_DATE_TIME_OPTS)
}

/** Coerce API creative (string | preset dict | null) to a display string. */
export function creativeDisplayLabel(creative: unknown): string | null {
  if (creative == null) return null
  if (typeof creative === 'string') {
    const trimmed = creative.trim()
    return trimmed || null
  }
  if (typeof creative === 'object' && !Array.isArray(creative)) {
    const obj = creative as Record<string, unknown>
    for (const key of ['sender_display_name', 'label', 'olc_template_name'] as const) {
      const v = obj[key]
      if (typeof v === 'string' && v.trim()) return v.trim()
    }
    const name = [obj.first_name, obj.last_name]
      .filter((p): p is string => typeof p === 'string' && Boolean(p.trim()))
      .join(' ')
      .trim()
    return name || null
  }
  const text = String(creative).trim()
  return text || null
}

function normalizeOne(entry: unknown, idx: number): MailerHistoryRow | null {
  if (entry == null || entry === '') return null

  if (typeof entry === 'object' && !Array.isArray(entry)) {
    const obj = entry as Record<string, unknown>
    const templateName = obj.template_name != null ? String(obj.template_name) : null
    const creative = creativeDisplayLabel(obj.creative)
    const labelParts = [templateName, creative]
      .map((p) => (typeof p === 'string' ? p.trim() : p))
      .filter(Boolean)
    let label = labelParts.length ? labelParts.join(', ') : null
    if (!label && obj.olc_silent_omit) label = 'OLC silent omit'
    if (!label && obj.olc_order_id) label = `OLC order ${obj.olc_order_id}`
    if (!label && obj.campaign_id != null) label = `Campaign ${obj.campaign_id}`
    if (!label && obj.address_feedback) label = `Address feedback: ${obj.address_feedback}`
    if (!label) label = 'Mailer'
    const rawSource = obj.source
    const source: MailerHistorySource =
      rawSource === 'timeline' || rawSource === 'olc' || rawSource === 'imported'
        ? rawSource
        : obj.campaign_id != null || obj.olc_order_id
          ? 'olc'
          : 'imported'
    return {
      id: `mail-${idx}`,
      sent_at: obj.sent_at != null ? String(obj.sent_at) : null,
      label,
      creative,
      template_name: templateName,
      campaign_id: typeof obj.campaign_id === 'number' ? obj.campaign_id : null,
      olc_order_id: obj.olc_order_id != null ? String(obj.olc_order_id) : null,
      address_feedback: obj.address_feedback != null ? String(obj.address_feedback) : null,
      address_failure_reason:
        obj.address_failure_reason != null ? String(obj.address_failure_reason) : null,
      olc_silent_omit: Boolean(obj.olc_silent_omit),
      cancelled: Boolean(obj.cancelled),
      source,
    }
  }

  const text = String(entry).trim()
  if (!text) return null
  const match = LEGACY_DATE_RE.exec(text)
  if (match?.groups) {
    return {
      id: `mail-${idx}`,
      sent_at: match.groups.date,
      label: match.groups.label.trim().replace(/,\s*$/, ''),
      creative: null,
      template_name: null,
      campaign_id: null,
      olc_order_id: null,
      address_feedback: null,
      address_failure_reason: null,
      olc_silent_omit: false,
      cancelled: false,
      source: 'imported',
    }
  }
  return {
    id: `mail-${idx}`,
    sent_at: null,
    label: text,
    creative: null,
    template_name: null,
    campaign_id: null,
    olc_order_id: null,
    address_feedback: null,
    address_failure_reason: null,
    olc_silent_omit: false,
    cancelled: false,
    source: 'imported',
  }
}

export function normalizeMailerHistory(raw: unknown): MailerHistoryRow[] {
  const rows: MailerHistoryRow[] = []
  asEntries(raw).forEach((entry, idx) => {
    const row = normalizeOne(entry, idx)
    if (row) rows.push(row)
  })
  return rows
}

export function mailerHistorySummary(raw: unknown): MailerHistorySummary {
  const rows = normalizeMailerHistory(raw)
  let lastSent: string | null = null
  let lastMs: number | null = null
  for (const row of rows) {
    if (!row.sent_at) continue
    const parsed = parseMailerSentAt(row.sent_at)
    if (parsed) {
      const ms = parsed.getTime()
      if (lastMs == null || ms > lastMs) {
        lastMs = ms
        lastSent = row.sent_at
      }
    } else if (lastSent == null) {
      lastSent = row.sent_at
    }
  }
  return { count: rows.length, last_sent_at: lastSent, rows }
}

/** Prefer API summary; fall back to client normalize of raw history. */
export function resolveMailerHistorySummary(
  summary: MailerHistorySummary | null | undefined,
  raw: unknown,
): MailerHistorySummary {
  if (summary && typeof summary.count === 'number' && Array.isArray(summary.rows)) {
    // API rows may still carry dict `creative` — coerce so React never gets an object child.
    return {
      ...summary,
      rows: summary.rows.map((row, idx) => {
        const creative = creativeDisplayLabel(row.creative)
        const templateName = row.template_name
        const labelFromParts = [templateName, creative].filter(Boolean).join(', ')
        const creativeWasNonString =
          row.creative != null && typeof row.creative !== 'string'
        const labelLooksLikeDict =
          typeof row.label === 'string'
          && /\{\s*['"]/.test(row.label)
        const preferRebuilt =
          (creativeWasNonString || labelLooksLikeDict) && Boolean(labelFromParts)
        return {
          ...row,
          id: row.id || `mail-${idx}`,
          creative,
          label: preferRebuilt ? labelFromParts : (row.label || labelFromParts || 'Mailer'),
          address_failure_reason: row.address_failure_reason ?? null,
          olc_silent_omit: Boolean(row.olc_silent_omit),
        }
      }),
    }
  }
  return mailerHistorySummary(raw)
}
